
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://Yichen-Si.github.io/punkst/input/">
      
      
        <link rel="prev" href="../basic/">
      
      
        <link rel="next" href="../modules/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.11">
    
    
      
        <title>Input - punkst</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.4af4bdda.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#notes-on-processing-data-from-different-technologiesplatforms" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="punkst" class="md-header__button md-logo" aria-label="punkst" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            punkst
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Input
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1M8 13h8v-2H8zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4a5 5 0 0 0 5-5 5 5 0 0 0-5-5"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="black" data-md-color-accent="indigo"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5M7 15a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/Yichen-Si/punkst" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    Yichen-Si/punkst
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../install/" class="md-tabs__link">
        
  
  
    
  
  Install

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../basic/" class="md-tabs__link">
        
  
  
    
  
  Quickstart

      </a>
    </li>
  

      
        
  
  
  
    
  
  
    <li class="md-tabs__item md-tabs__item--active">
      <a href="./" class="md-tabs__link">
        
  
  
    
  
  Input

      </a>
    </li>
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../modules/" class="md-tabs__link">
          
  
  
    
  
  Modules

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="punkst" class="md-nav__button md-logo" aria-label="punkst" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    punkst
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Yichen-Si/punkst" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    Yichen-Si/punkst
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../install/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Install
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../basic/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Quickstart
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Input
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Input
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#visium-hd" class="md-nav__link">
    <span class="md-ellipsis">
      Visium HD
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cosmx-smi" class="md-nav__link">
    <span class="md-ellipsis">
      CosMx SMI
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#merscope" class="md-nav__link">
    <span class="md-ellipsis">
      MERSCOPE
    </span>
  </a>
  
    <nav class="md-nav" aria-label="MERSCOPE">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#run-ficture-pipeline-from-raw-data" class="md-nav__link">
    <span class="md-ellipsis">
      Run FICTURE pipeline from raw data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#just-conver-the-data-format-to-generic-input" class="md-nav__link">
    <span class="md-ellipsis">
      Just conver the data format to generic input
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#xenium" class="md-nav__link">
    <span class="md-ellipsis">
      Xenium
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
          
        
      
        
      
        
      
        
      
        
      
        
      
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../modules/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    Modules
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5" id="__nav_5_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Modules
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../modules/pts2tiles/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    pts2tiles
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../modules/tiles2hex/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    tiles2hex
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../modules/lda4hex/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    lda4hex
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../modules/pixel-decode/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    pixel-decode
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../modules/visualization/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Visualization
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#visium-hd" class="md-nav__link">
    <span class="md-ellipsis">
      Visium HD
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cosmx-smi" class="md-nav__link">
    <span class="md-ellipsis">
      CosMx SMI
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#merscope" class="md-nav__link">
    <span class="md-ellipsis">
      MERSCOPE
    </span>
  </a>
  
    <nav class="md-nav" aria-label="MERSCOPE">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#run-ficture-pipeline-from-raw-data" class="md-nav__link">
    <span class="md-ellipsis">
      Run FICTURE pipeline from raw data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#just-conver-the-data-format-to-generic-input" class="md-nav__link">
    <span class="md-ellipsis">
      Just conver the data format to generic input
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#xenium" class="md-nav__link">
    <span class="md-ellipsis">
      Xenium
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="notes-on-processing-data-from-different-technologiesplatforms">Notes on processing data from different technologies/platforms<a class="headerlink" href="#notes-on-processing-data-from-different-technologiesplatforms" title="Permanent link">&para;</a></h1>
<h2 id="visium-hd">Visium HD<a class="headerlink" href="#visium-hd" title="Permanent link">&para;</a></h2>
<p>First we need to locate the "binned output" directory and the subdirectory with the original resolution. For the data downloaded from 10X website it is called <code>binned_outputs/square_002um</code>. Let's call it <code>RAWDIR</code>. In this directory, you should find the subdirecotries, <code>spatial</code> and <code>filtered_feature_bc_matrix</code> (I guess you could also use data in <code>raw_feature_bc_matrix</code> but I've not tried it yet).</p>
<p>In the <code>spatial</code> directory, you should have a json file that contains the scaling factor of the coordinates, named as <code>scalefactors_json.json</code>.
Let's grep the scaling factor (or set it manually).
<div class="language-bash highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="nv">microns_per_pixel</span><span class="o">=</span><span class="k">$(</span>grep<span class="w"> </span>-w<span class="w"> </span>microns_per_pixel<span class="w"> </span><span class="si">${</span><span class="nv">RAWDIR</span><span class="si">}</span>/spatial/scalefactors_json.json<span class="w"> </span><span class="p">|</span><span class="w"> </span>perl<span class="w"> </span>-lane<span class="w"> </span><span class="s1">&#39;$_ =~ m/.*&quot;microns_per_pixel&quot;: ([0-9.]+)/; print $1&#39;</span><span class="w"> </span><span class="k">)</span>
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="c1">#  &quot;microns_per_pixel&quot;: 0.2737129726047599 in an example data</span>
</span></code></pre></div></p>
<p>The spatial coordinates for each barcode are stored in <code>parguet</code> format, we do not support this format directly. Let's convert it to plain tsv file. <a href="https://duckdb.org/">duckdb</a> seems fast and easy to use:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a><span class="nb">cd</span><span class="w"> </span><span class="si">${</span><span class="nv">RAWDIR</span><span class="si">}</span>/spatial/
</span><span id="__span-1-2"><a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>duckdb<span class="w"> </span>-c<span class="w"> </span><span class="s2">&quot;COPY (SELECT * FROM read_parquet(&#39;tissue_positions.parquet&#39;)) TO &#39;tissue_positions.tsv&#39; (HEADER, DELIMITER &#39;\t&#39;);&quot;</span>
</span></code></pre></div>
<p>Alternatively, you can try to use <a href="https://pandas.pydata.org/docs/reference/api/pandas.read_parquet.html">pyarrow and pandas in python</a>.</p>
<p>Next, punkst has a command to merge the 10X style dge files (and the spatial coordinates) into a single file as our standard input:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-2-1"><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a><span class="nv">brc_raw</span><span class="o">=</span><span class="si">${</span><span class="nv">RAWDIR</span><span class="si">}</span>/spatial/tissue_positions.tsv<span class="w"> </span><span class="c1"># the one converted from parquet</span>
</span><span id="__span-2-2"><a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a><span class="nv">mtx_path</span><span class="o">=</span><span class="si">${</span><span class="nv">RAWDIR</span><span class="si">}</span>/filtered_feature_bc_matrix<span class="w"> </span><span class="c1"># path to the 10X style dge files</span>
</span><span id="__span-2-3"><a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a>punkst<span class="w"> </span>convert-dge<span class="w"> </span><span class="se">\</span>
</span><span id="__span-2-4"><a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a>--microns-per-pixel<span class="w"> </span><span class="si">${</span><span class="nv">microns_per_pixel</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
</span><span id="__span-2-5"><a id="__codelineno-2-5" name="__codelineno-2-5" href="#__codelineno-2-5"></a>--exclude-regex<span class="w"> </span><span class="s1">&#39;^mt-&#39;</span><span class="w"> </span>--in-tissue-only<span class="w"> </span><span class="se">\</span>
</span><span id="__span-2-6"><a id="__codelineno-2-6" name="__codelineno-2-6" href="#__codelineno-2-6"></a>--in-positions<span class="w"> </span><span class="si">${</span><span class="nv">brc_raw</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
</span><span id="__span-2-7"><a id="__codelineno-2-7" name="__codelineno-2-7" href="#__codelineno-2-7"></a>--in-dge-dir<span class="w"> </span><span class="si">${</span><span class="nv">mtx_path</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
</span><span id="__span-2-8"><a id="__codelineno-2-8" name="__codelineno-2-8" href="#__codelineno-2-8"></a>--output-dir<span class="w"> </span><span class="si">${</span><span class="nv">path</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
</span><span id="__span-2-9"><a id="__codelineno-2-9" name="__codelineno-2-9" href="#__codelineno-2-9"></a>--coords-precision<span class="w"> </span><span class="m">4</span>
</span></code></pre></div>
<p>Here the optional flag <code>--exclude-regex</code> takes a regular expression to exclude genes matching with the regex. In the above example we exclude all mitochondrial genes.</p>
<p>The optional flag <code>--in-tissue-only</code> will exclude all barcodes that are labeled as not in the tissue.</p>
<p>It wrote the main data <code>transcripts.tsv</code>, a file <code>coordinate_minmax.tsv</code> containing the coordinate range (xmin, xmax, ymin, ymax), and <code>features.tsv</code> listing the gene names and their total counts. We probably want to filter out some very rare probs/genes by <code>awk '$2 &gt; 100' ${path}/features.tsv &gt; ${path}/features_min100.tsv</code>. The coordinates in <code>transcripts.tsv</code> and the range in <code>coordinate_minmax.tsv</code> are all in microns.</p>
<p>Now you can use the standard Makefile and config file as described in <a href="basic.md">the quick start page</a> to run the pipeline (and specify all coordinate/size related parameters in microns).</p>
<h2 id="cosmx-smi">CosMx SMI<a class="headerlink" href="#cosmx-smi" title="Permanent link">&para;</a></h2>
<p>You can use the template <code>Makefile</code> and <code>config_prepare.json</code> in <code>punkst/examples/format_input/cosmx</code> to conver CosMx raw output files to the generic input format.
Copy the config file to your directory and set the raw file names. For example, here is an example for the public mouse half brain data:</p>
<div class="language-json highlight"><pre><span></span><code><span id="__span-3-1"><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a><span class="p">{</span>
</span><span id="__span-3-2"><a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a><span class="w">    </span><span class="nt">&quot;workflow&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">{</span>
</span><span id="__span-3-3"><a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a><span class="w">      </span><span class="nt">&quot;raw_tx&quot;</span><span class="w"> </span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;Run1000_S1_Half_tx_file.csv&quot;</span><span class="p">,</span>
</span><span id="__span-3-4"><a id="__codelineno-3-4" name="__codelineno-3-4" href="#__codelineno-3-4"></a><span class="w">      </span><span class="nt">&quot;raw_meta&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;Run1000_S1_Half_metadata_file.csv&quot;</span><span class="p">,</span>
</span><span id="__span-3-5"><a id="__codelineno-3-5" name="__codelineno-3-5" href="#__codelineno-3-5"></a><span class="w">      </span><span class="nt">&quot;raw_mtx&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;Run1000_S1_Half_exprMat_file.csv&quot;</span><span class="p">,</span>
</span><span id="__span-3-6"><a id="__codelineno-3-6" name="__codelineno-3-6" href="#__codelineno-3-6"></a><span class="w">      </span><span class="nt">&quot;microns_per_pixel&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">0.12</span><span class="p">,</span>
</span><span id="__span-3-7"><a id="__codelineno-3-7" name="__codelineno-3-7" href="#__codelineno-3-7"></a><span class="w">      </span><span class="nt">&quot;datadir&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;/output/test&quot;</span>
</span><span id="__span-3-8"><a id="__codelineno-3-8" name="__codelineno-3-8" href="#__codelineno-3-8"></a><span class="w">    </span><span class="p">}</span>
</span><span id="__span-3-9"><a id="__codelineno-3-9" name="__codelineno-3-9" href="#__codelineno-3-9"></a><span class="w">  </span><span class="p">}</span>
</span></code></pre></div>
<p>You can find <code>"microns_per_pixel"</code> in the ReadMe.html, it may say something like "To convert to microns multiply the pixel value by 0.12 um per pixel".</p>
<p>Then you can generate the concrete Makefile by
<div class="language-bash highlight"><pre><span></span><code><span id="__span-4-1"><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>python<span class="w"> </span>/path/to/punkst/ext/py/generate_workflow.py<span class="w"> </span><span class="se">\</span>
</span><span id="__span-4-2"><a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a><span class="w">  </span>-c<span class="w"> </span>config_prepare.json<span class="w"> </span>-m<span class="w"> </span>Makefile<span class="w"> </span><span class="se">\</span>
</span><span id="__span-4-3"><a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a><span class="w">  </span>-t<span class="w"> </span>/path/to/punkst/examples/format_input/cosmx/Makefile
</span></code></pre></div>
and run <code>make</code></p>
<h2 id="merscope">MERSCOPE<a class="headerlink" href="#merscope" title="Permanent link">&para;</a></h2>
<p>Template Makefile and config file to prepare standard input data from MERSCOPE outputs are in <code>punkst/examples/format_input/merscope</code>.</p>
<h3 id="run-ficture-pipeline-from-raw-data">Run FICTURE pipeline from raw data<a class="headerlink" href="#run-ficture-pipeline-from-raw-data" title="Permanent link">&para;</a></h3>
<p>Parameters in <code>config.json</code> (<code>examples/format_input/config.json</code>) are mostly the same as in <a href="basic.md">the generic example</a>. The only difference is that we start from the MERSCOPE output files.</p>
<ul>
<li><code>"rawdir"</code>: the path that contains the MERSCOPE output files. We will need the following files: <code>cell_by_gene.csv.gz</code>, <code>cell_metadata.csv.gz</code>, and <code>detected_transcripts.csv.gz</code>. If your data is compressed, set <code>"compressed"</code>
 to 1, otherwise (plain csv) set it to 0.</li>
</ul>
<p>If you use <code>examples/format_input/Makefile0</code>, we do not use any cell information. (In this case <code>cell_metadata.csv.gz</code> is only used to get the coordinate range since this file is not too big.)</p>
<p>If you use <code>examples/format_input/Makefile1</code>, it has an experimental feature to use cell centers in <code>cell_metadata.csv.gz</code> to guide model initialization. Sometimes it helps factorization to separate nuclei specific signatures from cell type signatures. The additional parameter <code>"nucleiradius"</code> is only used for this case: we stratify transcripts by their distance to cell center by this cutoff. (This value should be in the same unit as the coordinates in the raw data, so microns for MERSCOPE)</p>
<p>(You can generate a concrete Makefile by running <code>generate_workflow.py</code> as described in the <a href="basic.md">quick start page</a>.)</p>
<h3 id="just-conver-the-data-format-to-generic-input">Just conver the data format to generic input<a class="headerlink" href="#just-conver-the-data-format-to-generic-input" title="Permanent link">&para;</a></h3>
<p>If we separate out the data preparation step, there are only three parameters we need to set: <code>"rawdir"</code>, <code>"datadir"</code>, and <code>"compressed"</code> (see <code>examples/format_input/Makefile</code> and <code>config_prepare.json</code>). Executing Makefile created with this template is the same as executing <code>make prepare</code> for the full workflow Makefiles above.</p>
<h2 id="xenium">Xenium<a class="headerlink" href="#xenium" title="Permanent link">&para;</a></h2>
<p>Template Makefile and config files are in <code>punkst/examples/format_input/xenium</code>.</p>
<p>In the <code>config.json</code>, you need specify <code>"raw_transcripts"</code>, the path of the transcript file <code>transcripts.csv.gz</code>, and <code>"rawdir"</code> that contains the directory <code>cell_feature_matrix</code> (decompressed from <code>cell_feature_matrix.tar.gz</code>) and a cell metadata file <code>cells.csv.gz</code>.</p>
<p>There are two Makefiles to run FICTURE pipeline with (<code>Makefile0</code>) or without (<code>Makefile1</code>) cell center information, same as that for MERSCOPE (see above).</p>
<p>The following is what happens in the <code>prepare</code> rule in the Makefiles:</p>
<p>Basic on public datasets in <a href="https://www.10xgenomics.com/datasets">10X data release</a>, there will be a <code>transcripts.parquet.csv.gz</code> or <code>transcripts.csv.gz</code> file in the output bundle. We just need to decompress it and extract the columns we need: <code>x_location</code>, <code>y_location</code>, and <code>feature_name</code> (and add a dummy <code>count</code> as always 1).</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-5-1"><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>cat<span class="w"> </span>transcripts.parquet.csv.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-d<span class="s1">&#39;,&#39;</span><span class="w"> </span>-f<span class="w"> </span><span class="m">4</span>-6<span class="w"> </span><span class="p">|</span><span class="w"> </span>sed<span class="w"> </span><span class="s1">&#39;s/&quot;//g&#39;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>awk<span class="w"> </span>-F<span class="s1">&#39;,&#39;</span><span class="w"> </span>-v<span class="w"> </span><span class="nv">OFS</span><span class="o">=</span><span class="s2">$&quot;\t&quot;</span><span class="w"> </span><span class="s1">&#39; {print $2, $3, $1, &quot;1&quot;} &#39;</span><span class="w"> </span>&gt;<span class="w"> </span>transcripts.tsv
</span></code></pre></div>
<p>Decompress <code>cell_feature_matrix.tar.gz</code>, you can find the list of genes in <code>cell_feature_matrix/features.tsv.gz</code>. Unfortunately we need the first column contains the gene names in the transcript file, so let's decompress and extract the second column (excluding the negative control probes):
<div class="language-bash highlight"><pre><span></span><code><span id="__span-6-1"><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>zcat<span class="w"> </span>cell_feature_matrix/features.tsv.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span><span class="s2">&quot;Gene Expression&quot;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">2</span><span class="w"> </span>&gt;<span class="w"> </span>features.txt
</span></code></pre></div></p>
<p>If you don't know the range of the coordinates (needed in the standard Makefile workflow, though we only need that to make the final image), we could get it by parsing the cell metadata <code>cells.csv.gz</code> (just because it may be the smallest file with all coordinates):
<div class="language-bash highlight"><pre><span></span><code><span id="__span-7-1"><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>zcat<span class="w"> </span>cells.csv.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-d<span class="s1">&#39;,&#39;</span><span class="w"> </span>-f<span class="w"> </span><span class="m">2</span>-3<span class="w"> </span><span class="p">|</span><span class="w"> </span>tail<span class="w"> </span>-n<span class="w"> </span>+2<span class="w"> </span><span class="p">|</span><span class="w"> </span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">1000</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>awk<span class="w"> </span>-F<span class="s1">&#39;,&#39;</span><span class="w"> </span>-v<span class="w"> </span><span class="nv">OFS</span><span class="o">=</span><span class="s2">$&quot;\t&quot;</span><span class="w"> </span>-v<span class="w"> </span><span class="nv">out</span><span class="o">=</span><span class="s2">&quot;test_cell_centers.tsv&quot;</span><span class="w"> </span>-v<span class="w"> </span><span class="nv">range</span><span class="o">=</span><span class="s2">&quot;test_coord_range.txt&quot;</span><span class="w"> </span><span class="s1">&#39;\</span>
</span><span id="__span-7-2"><a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a><span class="s1">NR==1{xmin=$1; xmax=$1; ymin=$2; ymax=$2} \</span>
</span><span id="__span-7-3"><a id="__codelineno-7-3" name="__codelineno-7-3" href="#__codelineno-7-3"></a><span class="s1">{if ($1 &lt; xmin) {xmin=$1}; if ($1 &gt; xmax) {xmax=$1}; \</span>
</span><span id="__span-7-4"><a id="__codelineno-7-4" name="__codelineno-7-4" href="#__codelineno-7-4"></a><span class="s1">if ($2 &lt; ymin) {ymin=$2}; if ($2 &gt; ymax) {ymax=$2}; \</span>
</span><span id="__span-7-5"><a id="__codelineno-7-5" name="__codelineno-7-5" href="#__codelineno-7-5"></a><span class="s1">print $1, $2 &gt; out } \</span>
</span><span id="__span-7-6"><a id="__codelineno-7-6" name="__codelineno-7-6" href="#__codelineno-7-6"></a><span class="s1">END{print &quot;XMIN:=&quot; xmin &gt; range; \</span>
</span><span id="__span-7-7"><a id="__codelineno-7-7" name="__codelineno-7-7" href="#__codelineno-7-7"></a><span class="s1">print &quot;XMAX:=&quot; xmax &gt;&gt; range; \</span>
</span><span id="__span-7-8"><a id="__codelineno-7-8" name="__codelineno-7-8" href="#__codelineno-7-8"></a><span class="s1">print &quot;YMIN:=&quot; ymin &gt;&gt; range; \</span>
</span><span id="__span-7-9"><a id="__codelineno-7-9" name="__codelineno-7-9" href="#__codelineno-7-9"></a><span class="s1">print &quot;YMAX:=&quot; ymax &gt;&gt; range}&#39;</span>
</span></code></pre></div></p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../basic/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Quickstart">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                Quickstart
              </div>
            </div>
          </a>
        
        
          
          <a href="../modules/" class="md-footer__link md-footer__link--next" aria-label="Next: Overview">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                Overview
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": ["content.code.annotate", "content.code.copy", "content.tooltips", "navigation.expand", "navigation.footer", "navigation.indexes", "navigation.sections", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking"], "search": "../assets/javascripts/workers/search.f8cc74c7.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.c8b220af.min.js"></script>
      
    
  </body>
</html>